orchagent/vnetorch.cpp

```diff
@@ -848,6 +848,162 @@ bool VNetRouteOrch::removeNextHopGroup(const string& vnet, const NextHopGroupKey
+            {
+                setEndpointMonitor(vnet, monitors, nexthops_primary, monitoring, ipPrefix);
+            }
+            if (it_route->second.secondary != nexthops_secondary)
+            {
+                setEndpointMonitor(vnet, monitors, nexthops_secondary, monitoring, ipPrefix);
+            }
+            nexthops_selected = it_route->second.nhg_key;
+            return true;
+        }
+
+        NextHopGroupKey nhg_custom = createActiveNHSet( vnet, nexthops_primary, ipPrefix);
+        if (!hasNextHopGroup(vnet, nhg_custom))
+        {
+            if (!createNextHopGroup(vnet, nhg_custom, vrf_obj, monitoring))
+            {
+                SWSS_LOG_WARN("Failed to create Primary based custom next hop group. Cannot proceed.");
+                delEndpointMonitor(vnet, nexthops_primary, ipPrefix);
+                delEndpointMonitor(vnet, nexthops_secondary, ipPrefix);
+                monitor_info_[vnet].erase(ipPrefix);
+
+                return false;
+            }
+        }
+        if (nhg_custom.getSize() > 0 )
+        {
+            SWSS_LOG_INFO(" Created Primary based custom next hop group.%s", nhg_custom.to_string().c_str() );
+            nexthops_selected = nhg_custom;
+            return true;
+        }
+        NextHopGroupKey nhg_custom_sec = createActiveNHSet( vnet, nexthops_secondary, ipPrefix);
+        if (!hasNextHopGroup(vnet, nhg_custom_sec))
+        {
+            if (!createNextHopGroup(vnet, nhg_custom_sec, vrf_obj, monitoring))
+            {
+                SWSS_LOG_WARN("Failed to create secondary based custom next hop group. Cannot proceed.");
+                delEndpointMonitor(vnet, nexthops_primary, ipPrefix);
+                delEndpointMonitor(vnet, nexthops_secondary, ipPrefix);
+                monitor_info_[vnet].erase(ipPrefix);
+
+                return false;
+            }
+        }
+        if (nhg_custom_sec.getSize() > 0 )
+        {
+            SWSS_LOG_INFO(" Created Secondary based custom next hop group.(%s).", nhg_custom_sec.to_string().c_str() );
+            nexthops_selected = nhg_custom_sec;
+            return true;
+        }
+        // nhg_custom is empty. we shall create a dummy enpty NHG for book keeping.
+        if (!hasNextHopGroup(vnet, nhg_custom) && !hasNextHopGroup(vnet, nhg_custom_sec) )
+        {
+            NextHopGroupInfo next_hop_group_entry;
+            next_hop_group_entry.next_hop_group_id = SAI_NULL_OBJECT_ID;
+            next_hop_group_entry.ref_count = 0;
+            syncd_nexthop_groups_[vnet][nhg_custom] = next_hop_group_entry;
+        }
+        nexthops_selected = nhg_custom;
+        return true;
+    }
+    else if (!hasNextHopGroup(vnet, nexthops_primary))
+    {
+        SWSS_LOG_INFO("Creating next hop group  %s", nexthops_primary.to_string().c_str());
+        setEndpointMonitor(vnet, monitors, nexthops_primary, "", ipPrefix);
+        if (!createNextHopGroup(vnet, nexthops_primary, vrf_obj, ""))
```

Comment:
What happens when nexthops_secondary.getSize() == 0 and monitoring == "custom"? 

You will pass empty monitoring string to these functions..Will that work ok?

Reply:
That is an invalid configuration. it will be stopped early on. in Handle Tunnel